pkg/store/multitsdb.go

```diff
@@ -0,0 +1,284 @@
+	if logger == nil {
+		logger = log.NewNopLogger()
+	}
+	return &MultiTSDBStore{
+		logger:     logger,
+		component:  component,
+		tsdbStores: tsdbStores,
+	}
+}
+
+// Info returns store information about the Prometheus instance.
+func (s *MultiTSDBStore) Info(ctx context.Context, req *storepb.InfoRequest) (*storepb.InfoResponse, error) {
+	stores := s.tsdbStores()
+
+	resp := &storepb.InfoResponse{
+		StoreType: s.component.ToProto(),
+	}
+	if len(stores) == 0 {
+		return resp, nil
+	}
+
+	infos := make([]*storepb.InfoResponse, 0, len(stores))
+	for _, store := range stores {
+		info, err := store.Info(ctx, req)
+		if err != nil {
+			return nil, err
+		}
+		infos = append(infos, info)
+	}
+
+	resp.MinTime = infos[0].MinTime
+	resp.MaxTime = infos[0].MaxTime
+
+	for i := 1; i < len(infos); i++ {
+		if resp.MinTime > infos[i].MinTime {
+			resp.MinTime = infos[i].MinTime
+		}
+		if resp.MaxTime < infos[i].MaxTime {
+			resp.MaxTime = infos[i].MaxTime
+		}
+	}
+
+	// We can rely on every underlying TSDB to only have one labelset, so this
+	// will always allocate the correct length immediately.
+	resp.LabelSets = make([]storepb.LabelSet, 0, len(infos))
+	for _, info := range infos {
+		resp.LabelSets = append(resp.LabelSets, info.LabelSets...)
+	}
+
+	return resp, nil
+}
+
+type seriesSetServer struct {
+	grpc.ServerStream
+
+	ctx context.Context
+
+	warnCh warnSender
+	recv   chan *storepb.Series
+	cur    *storepb.Series
+
+	errMtx *sync.Mutex
+	err    error
+}
+
+func newSeriesSetServer(
+	ctx context.Context,
+	warnCh warnSender,
+) *seriesSetServer {
+	return &seriesSetServer{
+		ctx:    ctx,
+		warnCh: warnCh,
+		recv:   make(chan *storepb.Series),
+		errMtx: &sync.Mutex{},
+	}
+}
+
+func (s *seriesSetServer) Context() context.Context {
+	return s.ctx
+}
+
+func (s *seriesSetServer) Run(store *TSDBStore, r *storepb.SeriesRequest) {
+	err := store.Series(r, s)
+	if err != nil {
+		if r.PartialResponseDisabled {
+			s.errMtx.Lock()
+			s.err = err
+			s.errMtx.Unlock()
+		} else {
+			s.warnCh.send(storepb.NewWarnSeriesResponse(err))
+		}
+	}
+
+	close(s.recv)
+}
+
+func (s *seriesSetServer) Send(r *storepb.SeriesResponse) error {
+	series := r.GetSeries()
+	chunks := make([]storepb.AggrChunk, len(series.Chunks))
+	copy(chunks, series.Chunks)
+	s.recv <- &storepb.Series{
+		Labels: series.Labels,
+		Chunks: chunks,
+	}
+	return nil
+}
+
+func (s *seriesSetServer) Next() (ok bool) {
+	s.cur, ok = <-s.recv
+	return ok
+}
+
+func (s *seriesSetServer) At() ([]storepb.Label, []storepb.AggrChunk) {
+	if s.cur == nil {
+		return nil, nil
+	}
+	return s.cur.Labels, s.cur.Chunks
+}
+
+func (s *seriesSetServer) Err() error {
+	s.errMtx.Lock()
+	defer s.errMtx.Unlock()
+	return s.err
+}
+
+// Series returns all series for a requested time range and label matcher. The returned data may
+// exceed the requested time bounds.
+func (s *MultiTSDBStore) Series(r *storepb.SeriesRequest, srv storepb.Store_SeriesServer) error {
```

Comment:
This looks good to me, but ... it also looks similar to proxy? Why not using proxy itself?

Reply:
I implemented both, and the proxy implementation got very complex and felt wrong as I was faking stores to be real grpc servers. There are some similarities, especially in the Series function, but there are a number of assumptions that we can make because we are in a more tightly scoped scenario. I opted for this one as it ended up being the simpler implementation of the two.

Reply:
ack :+1: